integration/spark/src/main/scala/org/apache/spark/sql/CodeGenerateFactory.scala

```diff
@@ -0,0 +1,158 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.SparkContext
+import org.apache.spark.sql.catalyst.CatalystConf
+import org.apache.spark.sql.catalyst.optimizer.Optimizer
+import org.apache.spark.sql.catalyst.plans.logical.{Expand, LogicalPlan}
+import org.apache.spark.util.{ScalaCompilerUtil, Utils}
+
+class CodeGenerateFactory(version: String) {
+
+  val optimizerFactory = if (version.equals("1.6.2")) {
+    ScalaCompilerUtil.compiledCode(CodeTemplates.spark1_6_2_OptimizerString)
+      .asInstanceOf[AbstractCarbonOptimizerFactory]
+  } else {
+    ScalaCompilerUtil.compiledCode(CodeTemplates.defaultOptimizerString)
+      .asInstanceOf[AbstractCarbonOptimizerFactory]
+  }
+
+  val expandFactory = if (version.startsWith("1.5")) {
+    ScalaCompilerUtil.compiledCode(CodeTemplates.spark1_5ExpandString)
+      .asInstanceOf[AbstractCarbonExpandFactory]
+  } else {
+    new AbstractCarbonExpandFactory {
+      override def createExpand(expand: Expand, child: LogicalPlan): Expand = {
+        val loader = Utils.getContextOrSparkClassLoader
+        try {
+          val cons = loader.loadClass("org.apache.spark.sql.catalyst.plans.logical.Expand")
+            .getDeclaredConstructors
+          cons.head.setAccessible(true)
+          cons.head.newInstance(expand.projections, expand.output, child).asInstanceOf[Expand]
+        } catch {
+          case e: Exception => null
+        }
+      }
+    }
+  }
+
+  def createDefaultOptimizer(conf: CatalystConf, sc: SparkContext): Optimizer = {
+    val name = "org.apache.spark.sql.catalyst.optimizer.DefaultOptimizer"
+    val loader = Utils.getContextOrSparkClassLoader
+    try {
+      val cons = loader.loadClass(name + "$").getDeclaredConstructors
+      cons.head.setAccessible(true)
+      cons.head.newInstance().asInstanceOf[Optimizer]
+    } catch {
+      case e: Exception =>
+        loader.loadClass(name).getConstructor(classOf[CatalystConf])
+          .newInstance(conf).asInstanceOf[Optimizer]
+    }
+  }
+
+}
+
+object CodeGenerateFactory {
+
+  private var codeGenerateFactory: CodeGenerateFactory = _
+
+  def init(version: String): Unit = {
+    if (codeGenerateFactory == null) {
+      codeGenerateFactory = new CodeGenerateFactory(version)
+    }
+  }
+
+  def getInstance(): CodeGenerateFactory = {
+    codeGenerateFactory
+  }
+
+  def getInstance(version: String): CodeGenerateFactory = {
+    init(version)
+    codeGenerateFactory
+  }
+
+}
+
+object CodeTemplates {
+
+  val spark1_6_2_OptimizerString =
+    s"""
+       import org.apache.spark.sql._;
+       import org.apache.spark.sql.optimizer._;
+       import org.apache.spark.sql.catalyst.plans.logical._;
+       import org.apache.spark.sql.catalyst._;
+       import org.apache.spark.sql.catalyst.optimizer.Optimizer;
+
+       new AbstractCarbonOptimizerFactory {
+
+           override def createOptimizer(optimizer1: Optimizer, conf1: CarbonSQLConf): Optimizer = {
```

Comment:
suggest rename input parameter `optimizer1` to `optimizer`. the same in defaultOptimizerString.